### ISSUE TITLE
Incorrect narrowing in "and" expression with boolean in LHS

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#1142]
**Describe the bug**
Pyright reports that the type of a boolean "and" expression with a boolean in the LHS is the type of the value in the RHS rather than their union.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That `foo() and baz()` would be `bool | str` and not `str`.

**Screenshots or Code**
```python
# pyright: strict

def foo() -> bool:
    ...

def bar() -> int:
    ...

def baz() -> str:
    ...

reveal_type(foo() and bar())  # "int": ok
reveal_type(foo() and baz())  # "str": should be "bool | str" or "Literal[False] | str"
reveal_type(bar() and foo())  # "int | bool": ok; could also be "int" like "foo() and bar()"
reveal_type(baz() and foo())  # "str | bool": ok

```

**VS Code extension or command-line**
Either, 1.1.83

**Comment by erictraut** [comment]
Thanks for the bug report. This was a recent regression that I introduced when fixing another bug in the "and" and "or" evaluation code.

The proper types for your four examples are:
```python
reveal_type(foo() and bar())  # "int | Literal[False]"
reveal_type(foo() and baz())  # "str | Literal[False]"
reveal_type(bar() and foo())  # "int | bool"
reveal_type(baz() and foo())  # "str | bool"
```

Note that the third one cannot be just `int` because the value might evaluate to zero, which is considered falsy.

I've added your example above to our test suite (and augmented it with a few more conditions).

This will be fixed in the next release.

**Comment by layday** [comment]
> Note that the third one cannot be just `int` because the value might evaluate to zero, which is considered falsy.

I believe mypy resolves this to `int` on account of `bool` being a subclass of `int` so I thought Pyright might have been doing something similar.

> This will be fixed in the next release.

✨

**Comment by layday** [comment]
These two should be the other way around, no?

```py
class Foo:
    ...


class Bar:
    ...


def func2(a: Foo and Bar):
    t1: Literal["Foo"] = reveal_type(a)


def func2(a: Foo or Bar):
    t1: Literal["Bar"] = reveal_type(a)
```

Class instances are truthy.

**Comment by jakebailey** [comment]
`Foo and Bar` and `Foo or Bar` aren't really valid type annotations; those operators won't return the right thing. (Previously came up on microsoft/pylance-release#560.)

I think this is thread is about what you get when you or/and _instances_ of classes together, not the types themselves.

**Comment by jakebailey** [comment]
Hm, I guess you're right that those would evaluate to `Bar` and `Foo` respectively, but I can't imagine anyone actually does this...

**Comment by jakebailey** [comment]
> These two should be the other way around, no?

![image](https://user-images.githubusercontent.com/5341706/98300556-49cd7f80-1f6e-11eb-8e49-adf637e70ea0.png)

No, your snippet is backwards.

**Comment by layday** [comment]
The snippet is from Pyright's tests: https://github.com/microsoft/pyright/commit/d16dafd1faa97189ceae05333c0739c320fe583d.  But you are right, this isn't a valid type annotation anyway.

**Comment by jakebailey** [comment]
Ah! My mistake.

**Comment by erictraut** [comment]
@layday, good catch! I had the cases inverted in both the code and the test cases.

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.84, which I just published. It will also be included in the next release of Pylance.
